source/agora/script/Signature.d

```diff
@@ -74,23 +98,64 @@ align(1):
 *******************************************************************************/
 
 public string decodeSignature (const(ubyte)[] bytes,
-    out SigPair sig_pair) pure nothrow @safe @nogc
+    out SigPair sig_pair, out ulong pop_count) pure nothrow @safe @nogc
 {
-    if (bytes.length != SigPair.sizeof)
+    pop_count = 0;
+    if (bytes.length < Signature.sizeof + SigHash.sizeof)
         return "Encoded signature tuple is of the wrong size";
 
     const sig = toSignature(bytes[0 .. Signature.sizeof]);
     bytes.popFrontN(Signature.sizeof);
+    pop_count += Signature.sizeof;
 
-    assert(bytes.length == 1);
     const SigHash sig_hash = cast(SigHash)bytes[0];
     if (!isValidSigHash(sig_hash))
         return "Unknown SigHash";
+    bytes.popFrontN(SigHash.sizeof);
+    pop_count += SigHash.sizeof;
 
-    sig_pair = SigPair(sig, sig_hash);
+    ulong output_idx;
+    if (sig_hash & SigHash.Single && bytes.length >= ulong.sizeof)
+    {
+        output_idx = littleEndianToNative!ulong(bytes[0 .. ulong.sizeof]);
+        pop_count += ulong.sizeof;
+    }
+    else if (sig_hash & SigHash.Single)
+        return "Encoded signature does not have output idx";
+
+    sig_pair = SigPair(sig, sig_hash, output_idx);
     return null;
 }
 
+unittest
+{
+    SigPair pair = SigPair(Signature.init, SigHash.Single, 1);
+    auto serialized = pair[].dup;
+
+    SigPair decoded_pair;
+    ulong consumed;
+
+    assert(decodeSignature(serialized, decoded_pair, consumed) is null);
+    assert(pair == decoded_pair);
+    assert(consumed == serialized.length);
+
+    serialized ~= ubyte(44);
```

Comment:
Don't think you need the `ubyte()` here, the compiler knows the type of the LHS.